hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/FreqKSketch.java

```diff
@@ -0,0 +1,87 @@
+package org.hiero.sketch.spreadsheet;
+
+import org.hiero.sketch.dataset.api.ISketch;
+import org.hiero.sketch.dataset.api.Pair;
+import org.hiero.sketch.table.*;
+import org.hiero.sketch.table.api.IRowIterator;
+import org.hiero.sketch.table.api.ITable;
+import org.hiero.utils.Converters;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+public class FreqKSketch implements ISketch<ITable, FreqKList> {
+    private final RecordOrder recordOrder;
+    private final int maxSize;
+
+    public FreqKSketch(List<ColumnDescription> colDescList, int maxSize) {
+        this.recordOrder = new RecordOrder();
+        colDescList.forEach(cd -> this.recordOrder.append(new ColumnSortOrientation(cd, true)));
+        this.maxSize = maxSize;
+    }
+
+    @Nullable
+    @Override
+    public FreqKList zero() {
+        return new FreqKList(0, this.maxSize, new HashMap<>(0));
+    }
+
+    @Override
+    public FreqKList add(@Nullable FreqKList left, @Nullable FreqKList right) {
+        Converters.checkNull(left);
+        for (RowSnapshot rs : Converters.checkNull(right).hMap.keySet()) {
+            if (left.hMap.containsKey(rs)) {
+                left.hMap.put(rs, left.hMap.get(rs) + right.hMap.get(rs));
+            } else
+                left.hMap.put(rs, right.hMap.get(rs));
+        }
+        List<Pair<RowSnapshot, Integer>> pList = new ArrayList<>(left.hMap.size());
+        left.hMap.forEach((rs, j) -> pList.add(new Pair<RowSnapshot, Integer>(rs, j)));
+        pList.sort((p1, p2) -> Integer.compare(p2.second, p1.second));
+        int k = 0;
+        if(pList.size() >= (this.maxSize + 1)) {
+            k = pList.get(this.maxSize).second;
+        }
+        HashMap<RowSnapshot,Integer> hm = new HashMap<>(this.maxSize);
+        for (int i = 0; i < Math.min(this.maxSize, pList.size()); i++) {
+            if (pList.get(i).second >= (k + 1))
+                hm.put(pList.get(i).first, pList.get(i).second - k);
+        }
+        return new FreqKList(left.totalRows + right.totalRows, this.maxSize, hm);
+    }
+
+    @Override
+    public FreqKList create(ITable data) {
+        IRowIterator rowIt = data.getRowIterator();
+        Schema schema = this.recordOrder.toSchema();
+        HashMap <VirtualRowSnapshot, Integer> hMap = new HashMap<>(this.maxSize);
+        List <VirtualRowSnapshot> toRemove = new ArrayList<>(this.maxSize);
+        int i;
+        do {
```

Comment:
This may be better as a while loop instead of a do loop; then you can remove the if inside.